ambry-router/src/main/java/com/github/ambry/router/DeleteOperation.java

```diff
@@ -379,6 +414,60 @@ private void checkAndMaybeComplete() {
     }
   }
 
+  /**
+   * Try to replicate the blob and then retry the Delete Operation
+   * @return true if it's under retry. return false if cannot do retry or retry is finished.
+   */
+  private boolean retryWithReplicateBlob() {
+    // it's the first time we hit the 503
+    if (replicateBlobCallback == null) {
+      List<ReplicaId> successfulReplica = operationTracker.getSuccessReplica();
+      DataNodeId sourceDataNode = successfulReplica.get(0).getDataNodeId();
+      replicateBlobCallback = new ReplicateBlobCallback(blobId, sourceDataNode);
+
+      nonBlockingRouter.replicateBlob(blobId.getID(), this.getClass().getSimpleName(), sourceDataNode,
+          replicateBlobCallback);
+      logger.trace("Start the on-demand replication {} {}.", replicateBlobCallback.getBlobId(),
+          replicateBlobCallback.getSourceDataNode());
+      return true;
+    }
+
+    ReplicateBlobCallback.State state = replicateBlobCallback.getState();
+    switch (state) {
+      case REPLICATING_BLOB:
+        // Has sent out the ReplicateBlob operation and is waiting for its completion.
+        logger.trace("Running on-demand replication {} {}.", replicateBlobCallback.getBlobId(),
+            replicateBlobCallback.getSourceDataNode());
+        return true;
+      case REPLICATION_DONE:
+        // ReplicateBlob is done.
+        // If ReplicateBlob fails, recovery procedure is done. Will complete the DeleteOperation
+        if (replicateBlobCallback.getException() != null) {
+          logger.error("on-demand replication {} {} failed.", replicateBlobCallback.getBlobId(),
+              replicateBlobCallback.getSourceDataNode(), replicateBlobCallback.getException());
+          return false;
+        }
+        // If ReplicateBlob is successful, retry the DeleteOperation.
+        logger.info("On-demand replication {} {} is successful.", replicateBlobCallback.getBlobId(),
```

Comment:
trace ?

Reply:
On-demand replication will barely get triggered unless there is something wrong.
It's intended to be info.